### PR TITLE
feat: 2-3x inference speedup, faster than real-time

### PR DESCRIPTION
**Pull request by vatsalaggarwal**
dirty fix to speed up inference by porting gpt-fast... supports single utterance only, no batching...

First stage only:
RTX 4090: 230T/s (~1.5 seconds of speech generated in 1 second of wall-clock time)
H100: 382T/s (~2.5 seconds of speech generated in 1 second of wall-clock time)

These times are consistent across context lengths due to the static kv-cache used.

Notes:
- it can take a while to start up the first time (~1-3 minutes)
- due to compilation, restarting the script pays a penalty of ~20-50second

### CODE REVIEW

File: fam/llm/gptfast_model.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+        self.config = config
+
+        self.tok_embeddings = nn.Embedding(config.vocab_size, config.dim)
+        self.pos_embeddings = nn.Embedding(config.block_size, config.dim)
+        self.speaker_cond_pos = nn.Linear(config.speaker_emb_dim, config.dim, bias=False)
+        self.layers = nn.ModuleList(TransformerBlock(config) for _ in range(config.n_layer))
+        self.norm = RMSNorm(config.dim, eps=config.norm_eps)
+        self.output = nn.Linear(config.dim, config.vocab_size, bias=False)
+
+        self.freqs_cis: Optional[Tensor] = None
```
Comment (sidroopdaska) [inline fam/llm/gptfast_model.py:117]: remove

File: fam/llm/gptfast_model.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        self.n_local_heads = config.n_local_heads
+        self.dim = config.dim
+        self._register_load_state_dict_pre_hook(self.load_hook)
+
+    def load_hook(self, state_dict, prefix, *args):
+        if prefix + "wq.weight" in state_dict:
+            wq = state_dict.pop(prefix + "wq.weight")
+            wk = state_dict.pop(prefix + "wk.weight")
+            wv = state_dict.pop(prefix + "wv.weight")
+            state_dict[prefix + "wqkv.weight"] = torch.cat([wq, wk, wv])
```
Comment (sidroopdaska) [inline fam/llm/gptfast_model.py:194]: remove?

File: fam/llm/gptfast_sample_utils.py [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+
+torch._inductor.config.coordinate_descent_tuning = True
+torch._inductor.config.triton.unique_kernel_names = True
+torch._inductor.config.fx_graph_cache = (
+    True  # Experimental feature to reduce compilation times, will be on by default in future
+)
+
+
+import tiktoken
+from gptfast_model import Transformer
```
Comment (sidroopdaska) [inline fam/llm/gptfast_sample_utils.py:55]: comment on the ordering

File: fam/llm/gptfast_sample_utils.py [hunk 4]
```diff
@@ -0,0 +1,441 @@
...
+            k = k.replace(".mlp.swiglu.w3.", ".feed_forward.swiglu.w3.")
+        if ".ln_1." in k:
+            state_dict[k.replace(".ln_1.", ".attention_norm.")] = state_dict.pop(k)
+            k = k.replace(".ln_1.", ".attention_norm.")
+        if ".ln_2." in k:
+            state_dict[k.replace(".ln_2.", ".ffn_norm.")] = state_dict.pop(k)
+            k = k.replace(".ln_2.", ".ffn_norm.")
+        if ".mlp.c_proj." in k:
+            state_dict[k.replace(".mlp.c_proj.", ".feed_forward.w2.")] = state_dict.pop(k)
+            k = k.replace(".mlp.c_proj.", ".feed_forward.w2.")
```
Comment (sidroopdaska) [inline fam/llm/gptfast_sample_utils.py:293]: comment on whats happening here

File: fam/llm/gptfast_sample_utils.py [hunk 5]
```diff
@@ -0,0 +1,441 @@
...
+    print(f"Compilation time: {time.perf_counter() - t0:.2f} seconds")
+
+    return model, tokenizer, smodel, precision, model_size
+
+
+def main(
+    *,
+    model,
+    tokenizer,
+    smodel,
```
Comment (sidroopdaska) [inline fam/llm/gptfast_sample_utils.py:389]: remove stale args

File: fam/llm/gptfast_sample_utils.py [hunk 6]
```diff
@@ -0,0 +1,441 @@
...
+    precision,
+    model_size,
+    prompt: str,
+    spk_ref_path: str,
```
Comment (sidroopdaska) [inline fam/llm/gptfast_sample_utils.py:393]: remove stale args

File: fam/llm/gptfast_inference.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        # the torch.compiled-model.
+        self._model_dir = snapshot_download(repo_id="metavoiceio/metavoice-1B-v0.1")
+        self.first_stage_adapter = FlattenedInterleavedEncodec2Codebook(end_of_audio_token=1024)
+        second_stage_ckpt_path = f"{self._model_dir}/second_stage.pt"
+        config_second_stage = InferenceConfig(
+            ckpt_path=second_stage_ckpt_path,
+            num_samples=1,
+            seed=1337,
+            device="cuda",
+            dtype="bfloat16",
```
Comment (sidroopdaska) [inline fam/llm/gptfast_inference.py:35]: auto handle dtype. check fam/llm/utils.py

File: README.md [hunk 8]
```diff
@@ -70,6 +70,20 @@ python app.py
...
+
+0. Upgrade to pytorch 2.2.0: `pip install --upgrade torch torchvision torchaudio`
+1. Place voice reference samples in `fam/llm/assets` folder.
+2. Run inference in the following manner:
+```bash
+cd fam/llm
+python -i gptfast_inference.py
+>>> print(inferencer.synthesize("Hello world!", "assets/male.wav"))
+>>> print(inferencer.synthesize("Crazy fast speed coming right at you!", "assets/male.wav"))
+>>> print(inferencer.synthesize("Crazy fast speed coming right at you!", "assets/female.wav"))
```
Comment (sidroopdaska) [inline README.md:83]: tidy up, and add comments on when this should be used. integrate into serving.py

**Review by sidroopdaska — CHANGES_REQUESTED** [review]
few clarifications requested

File: README.md (hunk 8 above)
Comment (vatsalaggarwal) [reply README.md:83]: discussed offline - you'll do this tn

File: fam/llm/gptfast_inference.py (hunk 7 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_inference.py:35]: done

File: fam/llm/gptfast_model.py (hunk 1 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_model.py:117]: done

File: fam/llm/gptfast_model.py (hunk 2 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_model.py:194]: done

File: fam/llm/gptfast_sample_utils.py (hunk 3 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_sample_utils.py:55]: done

File: fam/llm/gptfast_sample_utils.py (hunk 4 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_sample_utils.py:293]: done

File: fam/llm/gptfast_sample_utils.py (hunk 5 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_sample_utils.py:389]: done

File: fam/llm/gptfast_sample_utils.py (hunk 6 above)
Comment (vatsalaggarwal) [reply fam/llm/gptfast_sample_utils.py:393]: done

**Review by sidroopdaska — APPROVED** [review]

**Pull request merged**